src/arch/arm/64/kernel/vspace.c

```diff
@@ -1664,26 +1664,25 @@ static exception_t decodeARMFrameInvocation(word_t invLabel, word_t length,
             return EXCEPTION_SYSCALL_ERROR;
         }
 #else
+        /* When in EL1, we are using the user page table for flushing and need to make sure
+           the mapping info in the cap is not stale. */
+        lookupPTSlot_ret_t lu_ret = lookupPTSlot(find_ret.vspace_root, vaddr);
+        pte_t pte = *lu_ret.ptSlot;
+        void *base_ptr = (void *) cap_frame_cap_get_capFBasePtr(cap);
+        if (unlikely(lu_ret.ptBitsLeft != pageBitsForSize(cap_frame_cap_get_capFSize(cap)) ||
+                     !pte_is_page_type(pte) ||
+                     pte_get_page_base_address(pte) != pptr_to_paddr(base_ptr))) {
+            userError("Page Flush: Attempting to use cap with stale mapping information.");
+            current_syscall_error.type = seL4_InvalidCapability;
+            current_syscall_error.invalidCapNumber = 0;
+            return EXCEPTION_SYSCALL_ERROR;
+        }
```

Comment:
Could this be lifted even higher to line: 1637, right after `findVSpaceForASID`?
In EL2, the kernel still shouldn't flush a cap with a stale ASID.

Reply:
In EL2, the operation is at least safe, because we do know the requested virtual address is mapped in the kernel window. I do agree that it would be better long-term to make this an error, because operating on stale mapping info is a user error.

Since moving it into EL2 will need re-verification I would like to merge this variant first, though, so we can release a version that is at least safe.